jobs: surface the trace_id linked to job execution

adityamaru opened the pull request:
Please see individual commit messages.

Informs: #64992

adityamaru committed:
jobs: save job trace ID in job payload

Previously, a job on creation would create a new context
and wrap it with a "forked span" from the parent ctx. This
meant that the span would inherit the traceID from its
remote parent. When the job began executing, we would create
yet another span. This span would be a child of the "forked
span" mentioned above, if the job was being started on the
same registry that created it. If the job was being resumed
by a new registry, we would create a root span.

This change deletes the root span created during job creation,
and always creates a root span when the job begins executing.
We did not use the span created on job creation for anything
and therefore this seems like an appropriate simplification.

Creating a root span allows us to identify only those spans linked
to job execution using the trace_id assigned to this root span.
This will be helpful when we build tooling to improve job
observability that will pull all the relevant spans.

This change also stores the trace_id associated with the current
job resumer in the jobs' payload. This allows us to use it at
a later time to inspect what a job is doing at a given point in
time!

Release note: None

adityamaru committed:
sql: surface the trace_id associated to a job

This change adds the trace_id associated with the
current execution of the job via the crdb_internal.jobs
table and as a column in `SHOW JOBS`. This will allow
users to pull the inflight traces for a job for debugging
purposes.

Release note (sql change): `SHOW JOBS` now shows the `trace_id`
of the trace that is associated with the current execution of the
job. This allows pulling inflight traces for a job for debugging
purposes.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/65322)
<!-- Reviewable:end -->

adityamaru commented:
@irfansharif adding you as a reviewer mainly for the first commit that changes the nature of the span associated with a job.

adityamaru commented:
Hmm the CI failure indicates that I have to `Finish()` the job root span somewhere to avoid leaking it. I'm not clear how we would previously `Finish()` the forked span, but the test cluster checks the node's registry before shutting down for any unfinished root spans. Please hold off on reviewing this till I figure this out!

adityamaru commented:
I think I just scared myself because of a unit test that was mocking out a job record and not going through the registry. We should not be leaking any spans. For posterity:

- The root span created when the job is created for the first time is marked as `Finished` once `runJob` returns in the `Start()` method.
- If the job is resumed on a node other than the one that created it, the root span created on resumption in `runJob` has a `defer span.Finish()`.

This is RFAL!

abarganier reviewed (commented):
Nice, LGTM. This seems like it will give some nice immediate returns! :sunglasses: I don't think I have enough context re: the new root spans being created in the 1st commit, but curious to hear @irfansharif's thoughts. 

> Reviewed 7 of 7 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65322#-:-M_uyGuq5SfMvOkQ48n3:bnhp4xp)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @irfansharif, and @pbardea)

___
*[pkg/jobs/adopt.go, line 258 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65322#-M_utCSQ4Yh6wZpnhzkJ:-M_utCSR4KfFIBzcXV9r:bv82mpz) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ad4586c02c8a4cf5317bc2abba7b83fe139be8f/pkg/jobs/adopt.go#L258)):*
> ```Go
> 	// If the job is being resumed by a registry, AnnotateCtxWithSpan will create
> 	// a new root span as ctx is not annotated with a parent span.
> 	ctx, span = r.ac.AnnotateCtxWithSpan(ctx, spanName, tracing.WithForceRealSpan())
> ```

More a curiosity question - can you link me to any resources that can help me understand the scenarios & semantics surrounding "resumed" jobs? 


<!-- Sent from Reviewable.io -->

irfansharif reviewed (commented):
Like discussed offline, I'd be curious to see what this PR would look like if we created a root span when only _running_ the job (both the very first time, and also during resumption), instead of creating the span when the job is first created, and not using it until executing it. That way, when running the job, we can unconditionally create a root span instead of "check if a span exists, if so use it, if not, create a new one because we're being resumed".

Also, from our offline discussions, the first commit message wants to be changed. Forked spans don't have to propagate payloads back to parent span. Everything else around storing the trace ID (and updating it on resumption) for future instropectability sounds sane to me. I'm excited for us to flesh out the primitives to read in-flight trace data from across the cluster, instead of the simpler primitive we have today that only renders a full recording of the trace's remote children only after it's done.

pbardea commented on pkg/jobs/registry.go:
```diff
@@ -530,15 +530,14 @@ func (r *Registry) CreateStartableJobWithTxn(
-			resumerCtx = tracing.ContextWithSpan(resumerCtx, span)
-		}
+		resumerCtx, span = r.settings.Tracer.StartSpanCtx(resumerCtx, "job",
+			tracing.WithForceRealSpan())
```
Along this line, one idea is to give jobs a verbosity level of their own that they can declare when creating their job record. That way particular types of jobs can opt-in to be traced (e.g. we're pretty confident there won't be a crazy amount of backup/restore jobs so those jobs can always enable tracing).

adityamaru replied:
This is a good idea, I haven't added it in yet because I want to flesh this out a little.

A span can either be a noop (this is the default unless tracing is on) or a "real" span. A noop span records nothing. A "real" span can further be in verbose mode or not. A non-verbose, real span allows for recording structured payloads but discards all free form logging. A verbose, real span allows for all forms of recording. In the initial google doc, knz had suggested we stick to structured payloads for tracing job execution.

So I'm inclined to allow jobs to opt-in to being a real, root span, but not allowing them to choose between verbose/non-verbose. What do you think?

pbardea replied:
That sounds good, my mention of jobs having a "verbosity" was to air on the side of making that flag an enum in the proto rather than a bool so that we can adjust later on if we want to add more fine-grained control.

adityamaru commented on pkg/jobs/adopt.go:
```diff
@@ -248,8 +248,23 @@ func (r *Registry) runJob(
 	defer span.Finish()
+	if err := job.Update(ctx, nil /* txn */, func(txn *kv.Txn, md JobMetadata,
+		ju *JobUpdater) error {
+		md.Payload.TraceId = span.TraceID()
```
nice catch, done.

adityamaru commented on pkg/jobs/jobs.go:
```diff
@@ -929,3 +921,10 @@ func (sj *StartableJob) Cancel(ctx context.Context) error {
+
+// FinishSpan will mark the span linked to the StartableJob as Finish()'d.
+func (sj *StartableJob) FinishSpan() {
+	if sj.span != nil {
```
makes sense, deleted the method though since we're no longer bookkeeping a span on job creation.

adityamaru commented:
> Like discussed offline, I'd be curious to see what this PR would look like if we created a root span when only _running_ the job (both the very first time, and also during resumption), instead of creating the span when the job is first created, and not using it until executing it. That way, when running the job, we can unconditionally create a root span instead of "check if a span exists, if so use it, if not, create a new one because we're being resumed".

Updated to only creating a root span when we start job execution. Also updated the first commit message to reflect this.

adityamaru commented:
> Nice, LGTM. This seems like it will give some nice immediate returns! 😎 I don't think I have enough context re: the new root spans being created in the 1st commit, but curious to hear @irfansharif's thoughts.
> 
> > Reviewed 7 of 7 files at r1.
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65322#-:-M_uyGuq5SfMvOkQ48n3:bnhp4xp)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @irfansharif, and @pbardea)
> 
> _[pkg/jobs/adopt.go, line 258 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65322#-M_utCSQ4Yh6wZpnhzkJ:-M_utCSR4KfFIBzcXV9r:bv82mpz) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ad4586c02c8a4cf5317bc2abba7b83fe139be8f/pkg/jobs/adopt.go#L258)):_
> 
> > ```go
> > 	// If the job is being resumed by a registry, AnnotateCtxWithSpan will create
> > 	// a new root span as ctx is not annotated with a parent span.
> > 	ctx, span = r.ac.AnnotateCtxWithSpan(ctx, spanName, tracing.WithForceRealSpan())
> > ```
> 
> More a curiosity question - can you link me to any resources that can help me understand the scenarios & semantics surrounding "resumed" jobs?

I'm not sure we have any docs around the specifics of job resumption. We have an old RFC on the jobs system though I think it has undergone fairly significant changes since then. In general, job creation and job execution are independent in the jobs infrastructure. A job when created is recorded once in the `system.jobs` table following which it is an acceptable candidate for adoption. Every node in the cluster runs a registry, and one of the processes in this registry polls the `system.jobs` table for adoptable jobs. When a registry adopts a job, it recreates the job from the information stored during creation, and begins execution. A job can subsequently be `PAUSED` and `RESUMED` multiple times, and this works by effectively toggling the adoptability of the job. There is no guarantee that the node that created the job/was previously running the job, will be the one to pick up the execution of the job on unpause.
A job might also be a candidate for resumption if the node running the job crashes. In this case, the current execution of the job will fail but will not mark the job entry as failed, thereby allowing other idle nodes to pickup its execution.

I'm happy to talk more about the job subsystem offline, if you have any questions!

miretskiy reviewed (commented):
> Reviewed 2 of 6 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65322)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier, @irfansharif, and @pbardea)

___
*[pkg/jobs/registry.go, line 540 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65322#-M_tntQ2CuqLGpgMhvKm-r1-540:-M_za8mJDrmDlWerB30J:b-3bfskk) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ad4586c02c8a4cf5317bc2abba7b83fe139be8f/pkg/jobs/registry.go#L540)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

This is a good idea, I haven't added it in yet because I want to flesh this out a little.

A span can either be a noop (this is the default unless tracing is on) or a "real" span. A noop span records nothing. A "real" span can further be in verbose mode or not. A non-verbose, real span allows for recording structured payloads but discards all free form logging. A verbose, real span allows for all forms of recording. In the initial google doc, knz had suggested we stick to structured payloads for tracing job execution.

So I'm inclined to allow jobs to opt-in to being a real, root span, but not allowing them to choose between verbose/non-verbose. What do you think?
</blockquote></details>

@adityamaru FYI: opting into being traced is easy.  I would just add a marker interface to jobs:
```
type TraceableJob interface {}
```

You can even add some methods there too (like: I want to be traced verbose or some such).
Have registry test if the resumer also implements this interface and create spans appropriately.


<!-- Sent from Reviewable.io -->

irfansharif commented on pkg/jobs/adopt.go:
```diff
@@ -248,8 +248,23 @@ func (r *Registry) runJob(
 	var span *tracing.Span
-	ctx, span = r.ac.AnnotateCtxWithSpan(ctx, spanName)
+
+	// Wrap the context with a root span that is tied to the execution of the
```
I'd drop the "Wrap the context" verbiage, lets just say "Create a new root span to trace the execution of the current instance of job...".

adityamaru replied:
done.

irfansharif commented on pkg/util/log/ambient_context.go:
```diff
@@ -158,7 +158,7 @@ func (ac *AmbientContext) annotateCtxInternal(ctx context.Context) context.Conte
 // The caller is responsible for closing the span (via Span.Finish).
 func (ac *AmbientContext) AnnotateCtxWithSpan(
-	ctx context.Context, opName string,
+	ctx context.Context, opName string, os ...tracing.SpanOption,
```
The diffs in these files can be reverted.

adityamaru replied:
done.

irfansharif reviewed (approved):
The use of pkg/tracing in this way LGTM.

pbardea reviewed (approved):
LGTM mod the discussion about opting into this.

Also, should there be a release note somewhere?

adityamaru reviewed (commented):
@pbardea added Yevgeniy's suggestion for opting in - https://github.com/cockroachdb/cockroach/pull/65322/files#diff-651e5b9db62dc86f9ef2a2d1937717a587117d43e9a589db3a43531817f2ca3bR257

In the future, we can add a `ForceVerboseSpan` if we deem it necessary. re: release note, I added one to the second commit about our new column in `SHOW JOBS`.

~edit: oops i think i broke all the tests with an NPE.~

abarganier reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65322)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier, @adityamaru, @irfansharif, and @miretskiy)

___
*[pkg/jobs/adopt.go, line 258 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65322#-M_utCSQ4Yh6wZpnhzkJ:-Ma-R6Pz3Hm6xJ-DJTDo:bfl1dtx) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ad4586c02c8a4cf5317bc2abba7b83fe139be8f/pkg/jobs/adopt.go#L258)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

I'm not sure we have any docs around the specifics of job resumption. We have an old RFC on the jobs system though I think it has undergone fairly significant changes since then. In general, job creation and job execution are independent in the jobs infrastructure. A job when created is recorded once in the `system.jobs` table following which it is an acceptable candidate for adoption. Every node in the cluster runs a registry, and one of the processes in this registry polls the `system.jobs` table for adoptable jobs. When a registry adopts a job, it recreates the job from the information stored during creation, and begins execution. A job can subsequently be `PAUSED` and `RESUMED` multiple times, and this works by effectively toggling the adoptability of the job. There is no guarantee that the node that created the job/was previously running the job, will be the one to pick up the execution of the job on unpause.
A job might also be a candidate for resumption if the node running the job crashes. In this case, the current execution of the job will fail but will not mark the job entry as failed, thereby allowing other idle nodes to pickup its execution.

I'm happy to talk more about the job subsystem offline, if you have any questions!

</blockquote></details>

Makes sense - thanks for the clear explanation, much appreciated 


<!-- Sent from Reviewable.io -->

adityamaru commented:
TFTRs!

bors r+

pbardea reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65322)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier, @adityamaru, @irfansharif, and @miretskiy)

___
*[pkg/jobs/adopt.go, line 268 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/65322#-Ma06DTkEyZj3eE1WBKu:-Ma06DTkEyZj3eE1WBKv:b-1khaea) ([raw file](https://github.com/cockroachdb/cockroach/blob/67a3392c358414fb71587177b34fef625ea9c8d7/pkg/jobs/adopt.go#L268)):*
> ```Go
> 		ju *JobUpdater) error {
> 		md.Payload.TraceID = span.TraceID()
> 		ju.UpdatePayload(md.Payload)
> ```

Should this be progress?

Update: as per offline discussion we'll change this in a follow-up
___
*[pkg/jobs/jobs.go, line 100 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/65322#-Ma05sLuBqziMCo8aI0U:-Ma05sLuBqziMCo8aI0V:b-doqdze) ([raw file](https://github.com/cockroachdb/cockroach/blob/67a3392c358414fb71587177b34fef625ea9c8d7/pkg/jobs/jobs.go#L100)):*
> ```Go
> 	// ForceRealSpan forces the registry to create a real Span instead of a
> 	// low-overhead non-recordable noop span.
> 	ForceRealSpan() bool
> ```

nit, take it or leave it: but it looks like every implementation of this is going to `return True`, so we can not even have a return value here and implementing it indicates that we want a real span/traceable job.


<!-- Sent from Reviewable.io -->

adityamaru commented:
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65322)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier, @adityamaru, @irfansharif, and @miretskiy)
> 
> _[pkg/jobs/adopt.go, line 268 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/65322#-Ma06DTkEyZj3eE1WBKu:-Ma06DTkEyZj3eE1WBKv:b-1khaea) ([raw file](https://github.com/cockroachdb/cockroach/blob/67a3392c358414fb71587177b34fef625ea9c8d7/pkg/jobs/adopt.go#L268)):_
> 
> > ```go
> > 		ju *JobUpdater) error {
> > 		md.Payload.TraceID = span.TraceID()
> > 		ju.UpdatePayload(md.Payload)
> > ```
> 
> Should this be progress?
> 
> Update: as per offline discussion we'll change this in a follow-up
> 
> _[pkg/jobs/jobs.go, line 100 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/65322#-Ma05sLuBqziMCo8aI0U:-Ma05sLuBqziMCo8aI0V:b-doqdze) ([raw file](https://github.com/cockroachdb/cockroach/blob/67a3392c358414fb71587177b34fef625ea9c8d7/pkg/jobs/jobs.go#L100)):_
> 
> > ```go
> > 	// ForceRealSpan forces the registry to create a real Span instead of a
> > 	// low-overhead non-recordable noop span.
> > 	ForceRealSpan() bool
> > ```
> 
> nit, take it or leave it: but it looks like every implementation of this is going to `return True`, so we can not even have a return value here and implementing it indicates that we want a real span/traceable job.

good point, addressed over in https://github.com/cockroachdb/cockroach/pull/65454.
